### PR TITLE
Update documentation to explain the behavior of explicitly raising ActiveRecord::Rollback in a nested transaction [skip ci]

### LINKED ISSUE #36229: Unexpected nested transaction rollback behavior with explicit ActiveRecord::Rollback
### Steps to reproduce
Raise an `ActiveRecord::Rollback` exception in a transaction (without a savepoint) that is nested in another one which has joinable set to true (the default).

### Expected behavior
The parent and child transaction should be rolled back.

### Actual behavior
Nothing is rolled back

### System configuration
**Rails version**: master, 5.2.2.1, (many more)
**Ruby version**: 2.5.5p157

### Reproduction Code

```ruby
# frozen_string_literal: true
# run with `ruby this_file.rb`

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title, null: false
  end
end

# Uncomment this module to apply patch to test how it affects test results.
#
# module ActiveRecordPatch
  # def transaction(requires_new: nil, isolation: nil, joinable: true)
    # if !requires_new && current_transaction.joinable?
      # if isolation
        # raise ActiveRecord::TransactionIsolationError, "cannot set isolation when joining a transaction"
      # end

      # yield
    # else
      # transaction_manager.within_new_transaction(isolation: isolation, joinable: joinable) { yield }
    # end
  # rescue ActiveRecord::Rollback
    # # NOTE: This is what was added to the original implementation
    # raise if !requires_new && current_transaction.joinable?
  # end
# end
# module ActiveRecord
  # module ConnectionAdapters # :nodoc:
    # class SQLite3Adapter
      # prepend ActiveRecordPatch
    # end
  # end
# end

class Post < ActiveRecord::Base
end

class MyError < StandardError
end

class BugTest < Minitest::Test
  def test_rollback_error
    ActiveRecord::Base.transaction do
      Post.create(title: 'first')
      ActiveRecord::Base.transaction do
        Post.create(title: 'second')
        raise ActiveRecord::Rollback
      end
      Post.create(title: 'last')
    end

    assert_nil(Post.find_by(title: 'first'))
    assert_nil(Post.find_by(title: 'second'))
    assert_nil(Post.find_by(title: 'last'))
  end
end
```

I've included a sample fix in the snippet above but I don't know enough about the AR internals to know if that's a reasonable solution to this issue.

### PR DESCRIPTION
**Pull request by vishaltelangre**
Closes https://github.com/rails/rails/issues/36229.

Reference: [ActiveRecord::Transactions](https://github.com/rails/rails/blob/6394f9da6938d0dccfd8c95f10b2fb7ee98ab953/activerecord/lib/active_record/transactions.rb#L135-L174).

This can be updated in master as well as all the previous applicable releases.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb [hunk 1]
```diff
@@ -198,26 +198,63 @@ def truncate_tables(*table_names) # :nodoc:
...
+      # effect of nested transactions, by using savepoints:
+      # https://dev.mysql.com/doc/refman/5.7/en/savepoint.html.
+      # Savepoints are supported by MySQL and PostgreSQL.
+      # SQLite3 version >= '3.6.8' supports savepoints.
       #
       # It is safe to call this method if a database transaction is already open,
-      # i.e. if #transaction is called within another #transaction block. In case
-      # of a nested call, #transaction will behave as follows:
+      # i.e. if #transaction is called within another #transaction block.
+      # In case of a nested call, #transaction will behave as follows:
```
Comment (vipulnsward) [inline activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb]: Lets revert these line wrap changes, they are not part of this change.

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb [hunk 2]
```diff
@@ -198,26 +198,63 @@ def truncate_tables(*table_names) # :nodoc:
...
-      # In order to get around this problem, #transaction will emulate the effect
-      # of nested transactions, by using savepoints:
-      # https://dev.mysql.com/doc/refman/5.7/en/savepoint.html
-      # Savepoints are supported by MySQL and PostgreSQL. SQLite3 version >= '3.6.8'
-      # supports savepoints.
+      # In order to get around this problem, #transaction will emulate the
+      # effect of nested transactions, by using savepoints:
+      # https://dev.mysql.com/doc/refman/5.7/en/savepoint.html.
+      # Savepoints are supported by MySQL and PostgreSQL.
+      # SQLite3 version >= '3.6.8' supports savepoints.
```
Comment (vipulnsward) [inline activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb]: Ditto

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb [hunk 3]
```diff
@@ -198,26 +198,63 @@ def truncate_tables(*table_names) # :nodoc:
...
+      #
+      #   ActiveRecord::Base.transaction do
+      #     Post.create(title: 'first')
+      #     ActiveRecord::Base.transaction do
+      #       Post.create(title: 'second')
+      #       raise ActiveRecord::Rollback
+      #     end
+      #   end
+      #
+      # creates both "first" and "second" posts. Reason is the
```
Comment (vipulnsward) [inline activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb]: This creates ...

**Comment by vishaltelangre** [comment]
@vipulnsward I have taken care of the suggested feedback. Please review.
